### PR TITLE
[WIP] Request manager spin loop and rerequest inefficiencies

### PR DESCRIPTION
**Pull request by gandrewstone**
The request manager had several recently introduced inefficiencies caused by moving the thin vs full block decision making into the request manager.  First, a communications cycle: a caller would request a block; the block request would request the block header first; receiving the block header would cause the block to be requested.  This cycle was fixed by not requesting the header if we already have it.  Second, if a call to RequestBlock resulted in no request, the system would push the same node back into the request manager as a data source, causing another no request -- repeatedly.  This spin loop resulted in a lot of logs.  This loop was fixed by leaving the no-request node in the request manager so subsequent AddSource requests for that same node just get ignored.  Next the request manager was changed to enable the normal request timeout if we decide to not request on that node, however, one optimization is special-cased: the first thin block capable node that is added as a source kicks the timeout to right now

### CODE REVIEW

**Review by ptschip — COMMENTED** [review]
while i can see that this will fix the problem of receiving headers over and over on regtest, i think this PR breaks some important functionlity: the thinblocktimer.  What I can see is going to happen without the thinblocktimer is we're going to request a lot more full blocks on mainnet. The thintimer is 10 seconds which was chosen as a fairly optimal number, the rerequest timeout is 5...now you're saying here the next time i re-request i'm going to ask for a full block.  I don't know how this plays out on mainnet, i haven't tested it but it's somewhat concerning and needs to be verified that it works...Secondly we still don't have a proper regression test for the requestManager...it's been on my list for some time but never get time to do it and this refactor here may be breaking other things just for the sake of fixing a regtest bug...I think this bug has been a nuisance for some time and again is on my list to finally root it out but can't get the time.  At the very least though I like this addition where you check if we already have the header before asking again.


        BlockMap::iterator idxIt = mapBlockIndex.find(obj.hash);
        pfrom->PushMessage(NetMsgType::GETHEADERS, chainActive.GetLocator(pindexBestHeader), obj.hash);
        if (idxIt == mapBlockIndex.end())  // only request if we don't already have the header
  	
Anyway, bottom line is, I think we should hold off on this PR until we get more testing done on mainnet and not rush this one in.
